- Exercici 1
Crea almenys dos models de classificació diferents per intentar predir el millor les classes de l'arxiu adjunt.

- Exercici 2
Compara els models de classificació utilitzant la precisió (accuracy), una matriu de confusió i d’altres mètriques més avançades.

- Exercici 3
Entrena’ls usant els diferents paràmetres que admeten per tal de millorar-ne la predicció.

- Exercici 4
Compara el seu rendiment fent servir l’aproximació traint/test o cross-validation.

- Exercici 5
Aplica algun procés d'enginyeria per millorar els resultats (normalització, estandardització, mostreig...)



### Previa:

- Carreguem les dades a analitzar (wineData.txt)
- El dataset proporciona les dades obtingudes dels analisis de vins d'una mateixa regió d'Italia procedents de 3 cultius diferents.
- Le dades obtingudes corresponen a 13 constituyents que es troben a cada un dels 3 tipus de vins.
- L'objectiu serà predir el tipus de vi en funció d'on ha estat cultivat. En funció d'on han estat cultivats es classifiquen en 3 clases -- 'Class': 1, 2 o 3

In [1]:
# Carreguem les dades a analitzar

import pandas as pd

pd.read_csv('wineData.txt')

# Afegim el nom de les columnes (constityents del vi) i afegim la columna 'Class' (cada un del cultius)

columnes = ['Class', 'Alcohol', 'Malic acid', 'Ash', 'Alcalinity of ash', 'Magnesium', 'Total phenols', 'Flavanoids',
            'Nonflavanoid phenols', 'Proanthocyanins', 'Color intensity', 'Hue', 'OD280/OD315 of diluted wines','Proline']

wine_df = pd.read_csv('wineData.txt', names = columnes)

pd.set_option('display.max_rows', None)

wine_df

Class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0        1    14.23        1.71  2.43               15.6        127   
1        1    13.20        1.78  2.14               11.2        100   
2        1    13.16        2.36  2.67               18.6        101   
3        1    14.37        1.95  2.50               16.8        113   
4        1    13.24        2.59  2.87               21.0        118   
5        1    14.20        1.76  2.45               15.2        112   
6        1    14.39        1.87  2.45               14.6         96   
7        1    14.06        2.15  2.61               17.6        121   
8        1    14.83        1.64  2.17               14.0         97   
9        1    13.86        1.35  2.27               16.0         98   
10       1    14.10        2.16  2.30               18.0        105   
11       1    14.12        1.48  2.32               16.8         95   
12       1    13.75        1.73  2.41               16.0         89   
13       1    14.75        1.73  2.39               11.4         91   
14       1    14.38        1.87  2.38               12.0        102   
15       1    13.63        1.81  2.70               17.2        112   
16       1    14.30        1.92  2.72               20.0        120   
17       1    13.83        1.57  2.62               20.0        115   
18       1    14.19        1.59  2.48               16.5        108   
19       1    13.64        3.10  2.56               15.2        116   
20       1    14.06        1.63  2.28               16.0        126   
21       1    12.93        3.80  2.65               18.6        102   
22       1    13.71        1.86  2.36               16.6        101   
23       1    12.85        1.60  2.52               17.8         95   
24       1    13.50        1.81  2.61               20.0         96   
25       1    13.05        2.05  3.22               25.0        124   
26       1    13.39        1.77  2.62               16.1         93   
27       1    13.30        1.72  2.14               17.0         94   
28       1    13.87        1.90  2.80               19.4        107   
29       1    14.02        1.68  2.21               16.0         96   
30       1    13.73        1.50  2.70               22.5        101   
31       1    13.58        1.66  2.36               19.1        106   
32       1    13.68        1.83  2.36               17.2        104   
33       1    13.76        1.53  2.70               19.5        132   
34       1    13.51        1.80  2.65               19.0        110   
35       1    13.48        1.81  2.41               20.5        100   
36       1    13.28        1.64  2.84               15.5        110   
37       1    13.05        1.65  2.55               18.0         98   
38       1    13.07        1.50  2.10               15.5         98   
39       1    14.22        3.99  2.51               13.2        128   
40       1    13.56        1.71  2.31               16.2        117   
41       1    13.41        3.84  2.12               18.8         90   
42       1    13.88        1.89  2.59               15.0        101   
43       1    13.24        3.98  2.29               17.5        103   
44       1    13.05        1.77  2.10               17.0        107   
45       1    14.21        4.04  2.44               18.9        111   
46       1    14.38        3.59  2.28               16.0        102   
47       1    13.90        1.68  2.12               16.0        101   
48       1    14.10        2.02  2.40               18.8        103   
49       1    13.94        1.73  2.27               17.4        108   
50       1    13.05        1.73  2.04               12.4         92   
51       1    13.83        1.65  2.60               17.2         94   
52       1    13.82        1.75  2.42               14.0        111   
53       1    13.77        1.90  2.68               17.1        115   
54       1    13.74        1.67  2.25               16.4        118   
55       1    13.56        1.

- Anàlisi previ del DataFrame 

In [2]:
wine_df.describe()

Class     Alcohol  Malic acid         Ash  Alcalinity of ash  \
count  178.000000  178.000000  178.000000  178.000000         178.000000   
mean     1.938202   13.000618    2.336348    2.366517          19.494944   
std      0.775035    0.811827    1.117146    0.274344           3.339564   
min      1.000000   11.030000    0.740000    1.360000          10.600000   
25%      1.000000   12.362500    1.602500    2.210000          17.200000   
50%      2.000000   13.050000    1.865000    2.360000          19.500000   
75%      3.000000   13.677500    3.082500    2.557500          21.500000   
max      3.000000   14.830000    5.800000    3.230000          30.000000   

        Magnesium  Total phenols  Flavanoids  Nonflavanoid phenols  \
count  178.000000     178.000000  178.000000            178.000000   
mean    99.741573       2.295112    2.029270              0.361854   
std     14.282484       0.625851    0.998859              0.124453   
min     70.000000       0.980000    0.340000              0.130000   
25%     88.000000       1.742500    1.205000              0.270000   
50%     98.000000       2.355000    2.135000              0.340000   
75%    107.000000       2.800000    2.875000              0.437500   
max    162.000000       3.880000    5.080000              0.660000   

       Proanthocyanins  Color intensity         Hue  \
count       178.000000       178.000000  178.000000   
mean          1.590899         5.058090    0.957449   
std           0.572359         2.318286    0.228572   
min           0.410000         1.280000    0.480000   
25%           1.250000         3.220000    0.782500   
50%           1.555000         4.690000    0.965000   
75%           1.950000         6.200000    1.120000   
max           3.580000        13.000000    1.710000   

       OD280/OD315 of diluted wines      Proline  
count                    178.000000   178.000000  
mean                       2.611685   746.893258  
std                        0.709990   314.907474  
min                        1.270000   278.000000  
25%                        1.937500   500.500000  
50%                        2.780000   673.500000  
75%                        3.170000   985.000000  
max                        4.000000  1680.000000

In [3]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Class                         178 non-null    int64  
 1   Alcohol                       178 non-null    float64
 2   Malic acid                    178 non-null    float64
 3   Ash                           178 non-null    float64
 4   Alcalinity of ash             178 non-null    float64
 5   Magnesium                     178 non-null    int64  
 6   Total phenols                 178 non-null    float64
 7   Flavanoids                    178 non-null    float64
 8   Nonflavanoid phenols          178 non-null    float64
 9   Proanthocyanins               178 non-null    float64
 10  Color intensity               178 non-null    float64
 11  Hue                           178 non-null    float64
 12  OD280/OD315 of diluted wines  178 non-null    float64
 13  Proli

- Fem la partició Train/Test del model

In [4]:
# Realitzem la partició separant el 'target' i la resta de variables (13 en total)

from sklearn.model_selection import train_test_split

x = wine_df.drop(['Class'], axis=1) # eliminem la variable 'Class' i mantenim la resta de variables 'Alchol', 'Malic Acid', etc.
y = wine_df['Class'] # Agafem només la variables 'Class'

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state =42) # Train = 80% / Test = 20%


In [5]:
# Estudiem Train / Test a nivell descriptiu

print("Descriptiu conjunt x_train:", x_train.describe())

print("Descriptiu conjunt y_train:", y_train.describe())

print("Descriptiu conjunt x_test:", x_test.describe())

print("Descriptiu conjunt y_train:", y_test.describe())

Descriptiu conjunt x_train:           Alcohol  Malic acid         Ash  Alcalinity of ash   Magnesium  \
count  142.000000  142.000000  142.000000         142.000000  142.000000   
mean    12.979085    2.373521    2.360845          19.473239  100.443662   
std      0.820116    1.143934    0.279217           3.454792   14.650793   
min     11.030000    0.890000    1.360000          10.600000   70.000000   
25%     12.332500    1.615000    2.210000          17.200000   88.250000   
50%     13.010000    1.875000    2.360000          19.200000   98.000000   
75%     13.677500    3.135000    2.540000          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
count     142.000000  142.000000            142.000000       142.000000   
mean        2.289085    2.002113              0.368028         1.608028   
std         0.637715    1.004170              0.128269        

-------------------------------------------------------------------------------------------------------------------------------

### Exercici 1

Crea almenys dos models de classificació diferents per intentar predir el millor les classes de l'arxiu adjunt.

#### - Arbre de Classificació (DecisionTreeClassifier)
   - L'objectiu de l'arbre de classificació és dividir el conjunt de dades en subconjuts mes petits i homogenis.
   - Cal controlar el sobreajust de les dades d'entrenament.
   - Revisar i valorar si cal reduïr la profunditat máxima de l'arbre.

In [6]:
from sklearn.tree import DecisionTreeClassifier

# Creem l'arbre de classificació:

DTC_wine = DecisionTreeClassifier()

# Arbre de classificació del model 'Train'

DTC_wine = DTC_wine.fit(x_train,y_train)

# Per predir la resposta per al model de 'Test'

pred_DTC_wine = DTC_wine.predict(x_test)

pred_DTC_wine

array([1, 1, 3, 1, 2, 1, 2, 3, 2, 3, 2, 1, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2,
       2, 3, 3, 3, 2, 2, 2, 1, 1, 2, 3, 1, 1, 1], dtype=int64)

--------------------------------------------------------------------------------------------------------------------------------

#### - SVM: Support Vector Machines
   - S'utilitza per classificar les dades en dues o més clases
   - Divideix l'espai de les dades en dues parts utiltzant un hiperplà.
   - El 'marge' es la distancia entre els punts mes propers de les dues classes.
   - L'objectiu del SVM es torbar un hiperplà amb el major marge possible.

In [7]:
from sklearn import svm

# Creem el model SVM

SVM_wine = svm.SVC() # kernel

# Entrenem el model 'Train'

SVM_wine.fit(x_train, y_train)

# Per predir la resposta per al model de 'Test

pred_SVM_wine = SVM_wine.predict(x_test)

pred_SVM_wine

array([1, 1, 2, 1, 2, 1, 2, 3, 3, 3, 1, 2, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2,
       3, 3, 2, 2, 2, 3, 2, 1, 1, 2, 3, 1, 1, 1], dtype=int64)

--------------------------------------------------------------------------------------------------------------------------------

### Exercici 2

Compara els models de classificació utilitzant la precisió (accuracy), una matriu de confusió i d’altres mètriques més avançades.

 ##### - Classification Acuracy: percentatge de prediccions correctes

In [8]:
from sklearn import metrics

accuracy_DTC_wine = metrics.accuracy_score(y_test, pred_DTC_wine)
accuracy_SVM_wine = metrics.accuracy_score(y_test, pred_SVM_wine)

# Realitzem un DF per visualitzar els resultats

accuracy_df = pd.DataFrame(
    data = [('Decision Tree Classifier',accuracy_DTC_wine), ('Support Vector Machines',accuracy_SVM_wine)],
        columns = ['Model', 'Accuracy'])

accuracy_df

Model  Accuracy
0  Decision Tree Classifier  0.944444
1   Support Vector Machines  0.805556

Comentaris:

- Els model DTC té un alt percentatge de prediccions correctes (94%), mentre que el model SVM te un accuracy sensiblement inferior (80%)
- En aquest cas, el model 'Decision Trees Classifier' sería més precís que el SVM.

 ##### - Confusion Matrix

In [9]:
from sklearn import metrics

cm_DTC_wine = metrics.confusion_matrix(y_test, pred_DTC_wine)

cm_DTC_wine

array([[13,  1,  0],
       [ 0, 14,  0],
       [ 1,  0,  7]], dtype=int64)

In [10]:
cm_SVM_wine = metrics.confusion_matrix(y_test, pred_SVM_wine)

cm_SVM_wine

array([[14,  0,  0],
       [ 0, 11,  3],
       [ 0,  4,  4]], dtype=int64)

In [11]:
# Pasem la Confusion Matrix del model DecisionTreeClassifier a DataFrame

cm_DTC_wine_DF = pd.DataFrame(cm_DTC_wine, index = ['Class 1', 'Class 2', 'Class 3'], columns = ['Pred.DTC Class 1', 'Pred.DTC Class 2', 'Pred.DTC Class 3'])

cm_DTC_wine_DF
                              

Pred.DTC Class 1  Pred.DTC Class 2  Pred.DTC Class 3
Class 1                13                 1                 0
Class 2                 0                14                 0
Class 3                 1                 0                 7

In [12]:
# Pasem la Confusion Matrix del model SVM a DataFrame

cm_SVM_wine_DF = pd.DataFrame(cm_SVM_wine, index = ['Class 1', 'Class 2', 'Class 3'], columns = ['Pred.SVM Class 1', 'Pred.SVM Class 2', 'Pred.SVM Class 3'])

cm_SVM_wine_DF
                        


Pred.SVM Class 1  Pred.SVM Class 2  Pred.SVM Class 3
Class 1                14                 0                 0
Class 2                 0                11                 3
Class 3                 0                 4                 4

Comentaris:

- Els model DTC (DecisionTreeClassfier) ha tingut errors en la predicció de la Class 1 i la Class 3, acomulant un total de 2 errors.
- Els model SVM (Support Vector Machines) ha tingut errors en la predicció de la Class 2 i la Class 3, acomulant un total de 7 errors.
- Podriem dir que, com en el cas de 'Accuracy', el model DTC sembla més precís que el model SVM.

 ##### - Logistic Loss

In [13]:
from sklearn.metrics import log_loss

# Calculem logistic loss del model DTC (DecisionTreeModel)

Loss_DTC_wine = log_loss(y_test, DTC_wine.predict_proba(x_test))

# Calculem logistic loss del model SVM (Support Vector Machines)

SVM_wine = svm.SVC(probability=True)

SVM_wine.fit(x_train, y_train)

Loss_SVM_wine = log_loss(y_test, SVM_wine.predict_proba(x_test))

print('Logistic loss DTC:', Loss_DTC_wine)
print('Logistic loss SVM:', Loss_SVM_wine)

Logistic loss DTC: 1.9188209108283725
Logistic loss SVM: 0.4801887582914907


Comentaris:

- Amb 'log_loss' es tracta d'obtenir la probabilitat de la similitud entre els valors predictius (probabilitats) i els valors reals.
- Utilitzem 'predict_proba' per obtenir les probabilitats de predicció del model.
- A menor valor de 'log_loss', millor serán les prediccions del model.
- En aques cas, podríem dir que les prediccions del model SVM podríen ser millors respecte a les del model DTC.

##### - Clasification Report

- Precision: Percentatge de prediccions positives correctes respecte el total de prediccions positives.

- Recall: Percentatge de prediccions positives correctes en relació amb el total de positius reals.

- F1 Score: Mitjana harmònica ponderada de 'Precision' i ' Recall'. A més proximitat al 1, millor és el model.

In [14]:
# Clasificatcion Report del model DTC (DecisionTreeClassifier)

print(metrics.classification_report(y_test, pred_DTC_wine, digits=3))

              precision    recall  f1-score   support

           1      0.929     0.929     0.929        14
           2      0.933     1.000     0.966        14
           3      1.000     0.875     0.933         8

    accuracy                          0.944        36
   macro avg      0.954     0.935     0.942        36
weighted avg      0.946     0.944     0.944        36



In [15]:
# Clasificatcion Report del model SVM (Support Vector Machines)

print(metrics.classification_report(y_test, pred_SVM_wine, digits=3))

              precision    recall  f1-score   support

           1      1.000     1.000     1.000        14
           2      0.733     0.786     0.759        14
           3      0.571     0.500     0.533         8

    accuracy                          0.806        36
   macro avg      0.768     0.762     0.764        36
weighted avg      0.801     0.806     0.802        36



Comentaris:

- Amb el 'Clasification Report' obtenim les mètriques de classificació del nostres model.
- Precision:
    - 'DecissionTreeClassifier':
        - la precisió per a la classe 1 és del 93%, indica que algunes mostres classificades com a classe 1 son en realitat d'una altra classe. Passa el mateix amb la classe 2, amb una precisió del 93%. Per a la classe 3, la precisió és del 100%, el que indica que totes les mostres classificades com a classe 3 son realment d'aquesta classe.
    - 'Support Vector Machines':
        - la precisió per a la classe 1 és del 100%, totes les mostres classificades com a classe 1 son realment d'aquesta classe. La precisió de la classe 2, és del 73% el que indica que algunes mostres classificades com a classe 2 son en realitat d'una altra classe. Per a la classe 3, la precisió és del 57%, el que indica que moltes de les mostres classificades com a classe 3 son en d'una altra classe. 
- Recall (Sensibilitat):
    - 'DecissionTreeClassifier':
        - El recall per a la classe 1 és del 93%, el que significa que algunes mostres de la classe 1 s'han classificat incorrectament com a una altra classe. Per a la classe 2, el recall és del 100%, el que indica que totes les mostres de la classe 2 s'han classificat correctament. La classe 3, el recall és del 87,5%, com en el cas de la classe 1 algunes mostres s'han classificat incorrectament.
    - 'Support Vector Machines':
         - El recall per a la classe 1 és del 100%, el que indica que totes les mostres de la classe 1 s'han classificat correctament. Per a la classe 2, el recall és del 79%, indica que algunes mostres de la classe 2 s'han classificat incorrectament com a una altra classe. La classe 3, el recall és del 50%, moltes de les mostres s'han classificat incorrectament.
- Accuracy: mesura la proporció de mostres classificades correctament.
    - 'DecissionTreeClassifier':
         - l'Accuracy global del model és del 94%. Podríem dir que el model té un bon rendiment en la classificació de les 3 classes.
    - 'Support Vector Machines':
         - En aquest cas, el valor és del 80%. Aquest model pdoríem dir que no presenta tant bon rendiment en la classificació de les 3 classes, respecte al model de DecissionTreeClassifier.
        




--------------------------------------------------------------------------------------------------------------------------------

### Exercici 3
   Entrena’ls usant els diferents paràmetres que admeten per tal de millorar-ne la predicció.

##### 1- DecisionTreeClassifier

- Inicialment utilitzem GridSearchCV per trobar el millors paràmetres del model.

In [16]:
from sklearn.model_selection import GridSearchCV

# Definim el model

DTC_wine #utilitzem el model anteriorment creat (arbre de classificació)

# Definim la llista de hierparamtres a ajustar i els valors a provar:

param_grid = {'max_depth': [2,4,6,8,10], 'criterion': ['gini', 'antropy'], 'min_samples_split': [2,4,6,8,10]}

# CrossValidation 

cv = 5

    #'max_depth': Profunditat máxima de l'arbre. L'arbre es dividirà fins el valor màxim definit o fins que les fulles siguin pures.
                    # Limitar la profunditat màxima pot ajudar a prevenir el sobreajust del model.

    #'Criterion': Criteris disponibles per als arbres de decissió (impuressa i incertessa del conjunt de dades)
                # 'Gini': La impuresa de 'Gini' es defineix com la probabilitat de que dos elements seleccionats aleatoriament d'un conjunt de dades siguin de diferents classes.
                            # Un valor de 'Gini' proper a '0' indica que el conjunt de dades és més homogeni.
                # 'Antropy': Es defineix com la mesura d'incertessa d'una variable aleatoria.En arbres de decissió s'utilitza l'antropia per medir la quantitat d'informació que es guanya després de realitzar la divisió.
                            # A subconjunts més homogenis, antropia menor. A subconjunt més heterogenis, antropia major.
    #'mini_sample_split': nº mínim de mostres que es necessiten per dividir un node intern de l'arbre.
    
    #'CV' (CrossValidation): permet evaluar el rendiement del model i evitar el sobreajustament.
                            # Especificant el valor de CV, fixem el número de 'folds'. Iteracions realitzades en el model
    
    # Els parametres 'max_depht' i 'mini_sample_split' son importants en la construcció d'un model d'arbre de decissió òptim.
    # S'han d'ajustar mitjançant validació creuada.

# Definim la GridSearch

grid_search_DTC_wine = GridSearchCV(DTC_wine, param_grid, cv=cv)

# Entrenem el model amb les dades d'entrenament

grid_search_DTC_wine.fit(x_train,y_train)

# Obtenim els millors parametres i el 'score' del model

best_params = grid_search_DTC_wine.best_params_
best_score = grid_search_DTC_wine.best_score_

print("Els millors hiperparàmtres son:", best_params)
print("El 'score' amb els millors hiperparàmetres és:", best_score)



Els millors hiperparàmtres son: {'criterion': 'gini', 'max_depth': 4, 'min_samples_split': 2}
El 'score' amb els millors hiperparàmetres és: 0.9224137931034484


C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
125 fits failed out of a total of 250.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
125 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 352, in fit
    criterion = CRITERIA_CLF[self.criterion](
KeyError: 'antropy'

  warnings.warn(some_

- Fem el nou model utilitzant els millors hiperparàmtres obtinguts i l'entrenem.

In [17]:
# Generem el nou model i l'entrenem

DTC_wine_2 = DecisionTreeClassifier(criterion= 'gini', max_depth = 4, min_samples_split= 2)
DTC_wine_2.fit(x_train,y_train)
pred_DTC_wine_2 = DTC_wine_2.predict(x_test)


In [18]:
# Calculem la 'Accuracy', la 'ConfusionMatrix' i 'LogisticLoss' del nou model per poder comparar-ho amb el model incial

accuracy_DTC_wine_2 = metrics.accuracy_score(y_test, pred_DTC_wine_2)

cm_DTC_wine_2 = metrics.confusion_matrix(y_test, pred_DTC_wine_2)

Loss_DTC_wine_2 = log_loss(y_test, DTC_wine_2.predict_proba(x_test))


In [19]:
# Generem un DF per visualitzar millor les dades a comparar dels dos models, inicial i entrenat amb paràmetres (Accuracy + Log_Loss)

DTC_wine_comp = pd.DataFrame( data = [('DTC Incicial', accuracy_DTC_wine, Loss_DTC_wine), 
                                  ('DTC Paràmetres', accuracy_DTC_wine_2, Loss_DTC_wine_2)],
                         columns = ['Model', 'Accuracy', 'Log_Loss'])
DTC_wine_comp


Model  Accuracy  Log_Loss
0    DTC Incicial  0.944444  1.918821
1  DTC Paràmetres  0.944444  1.918821

In [20]:
# Generem un DF de la confusion matrix del model entrenat amb paràmetres

cm_DTC_wine_2_DF = pd.DataFrame(cm_DTC_wine_2, index = ['Class 1', 'Class 2', 'Class 3'],
                                 columns = ['Pred.DTC_2 Class 1', 'Pred.DTC_2 Class 2', 'Pred.DTC_2 Class 3'])

print('ConfusionMatrix model DTC entrenat amb paràmetres:')

cm_DTC_wine_2_DF


ConfusionMatrix model DTC entrenat amb paràmetres:


Pred.DTC_2 Class 1  Pred.DTC_2 Class 2  Pred.DTC_2 Class 3
Class 1                  13                   1                   0
Class 2                   0                  14                   0
Class 3                   0                   1                   7

In [21]:
# i la comparem amb la del model incial.

print('ConfusionMatrix model DTC inicial:')

cm_DTC_wine_DF


ConfusionMatrix model DTC inicial:


Pred.DTC Class 1  Pred.DTC Class 2  Pred.DTC Class 3
Class 1                13                 1                 0
Class 2                 0                14                 0
Class 3                 1                 0                 7

--------------------------------------------------------------------------------------------------------------------------------

##### 2- SVM: Support Vector Machines

- Inicialment utilitzem GridSearchCV per trobar el millors paràmetres del model.

In [22]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

# Definim el model

SVM_wine #utilitzem el model anteriorment creat (arbre de classificació)

# Definim la llista de hierparamtres a ajustar i els valors a provar:

param_grid_SVM_wine = {'C': [0.01, 0.1, 1, 10, 100],'gamma': ['scale', 'auto'], 'degree': [1,2,3,4,5],'probability': [True]}

cv = 5

    #'C': Paràmtre de regulació de SVM, que controla la compensació (trade-off) entre la precissió en la classificació de les dades d'entrenament i la simplicitat del model. 
                    # Un valor mes alt de 'C', incrementa la precisió en la classificació de les dades d'entrenament, però pot dur a un sobreajustament del model.
    #'Gamma': Defineix la influencia de cada punt d'entrenament en la definició del límit de decissió.
                    # Un valor mes alt de 'gamma' fa que el model sigui mes sensible a punts propers, però pot dur a sobreajustament.
                        # 'scale': 1 / (n_features * X.var()) Escala les característiques a mitjana '0' i variança unitaria
                        # ' auto': 1 / n_features
    #'degree': Defineix el grau del polinomi utilitzat per transformar les dades d'entrada en un espai de característiques de major dimensió.
                    # Un valor mes alt de 'degree' pot incremnetar la complexitat del model.
    #'probability': habilitant aquesta opció, en lloc de simplement predir la classe de la mostra d'entrada, el model també pot estimar la probabilitat de que la mostra pertanyi a cada classe.
    
    #'CV' (CrossValidation): permet evaluar el rendiement del model i evitar el sobreajustament.
                            # Especificant el valor de CV, fixem el número de 'folds'. Iteracions realitzades en el model
    
# Definim la GridSearch

grid_search_SVM_wine = GridSearchCV(SVM_wine, param_grid_SVM_wine, cv=cv)

# Entrenem el model amb les dades d'entrenament

grid_search_SVM_wine.fit(x_train,y_train)

# Obtenim els millors parametres i el 'score' del model

best_params = grid_search_SVM_wine.best_params_
best_score = grid_search_SVM_wine.best_score_

print("Els millors hiperparàmtres son:", best_params)
print("El 'score' amb els millors hiperparàmetres és:", best_score)



Els millors hiperparàmtres son: {'C': 100, 'degree': 1, 'gamma': 'scale', 'probability': True}
El 'score' amb els millors hiperparàmetres és: 0.7187192118226602


- Fem el nou model utilitzant els millors hiperparàmtres obtinguts i l'entrenem.

In [23]:
# Generem el nou model i l'entrenem

SVM_wine_2 = svm.SVC(C = 100, degree = 1, gamma = 'scale', probability = True)
SVM_wine_2.fit(x_train,y_train)
pred_SVM_wine_2 = SVM_wine_2.predict(x_test)

In [24]:
# Calculem la 'Accuracy', la 'ConfusionMatrix' i 'LogisticLoss' del nou model per poder comparar-ho amb el model incial

accuracy_SVM_wine_2 = metrics.accuracy_score(y_test, pred_SVM_wine_2)

cm_SVM_wine_2 = metrics.confusion_matrix(y_test, pred_SVM_wine_2)

Loss_SVM_wine_2 = log_loss(y_test, SVM_wine_2.predict_proba(x_test))

In [25]:
# Generem un DF per visualitzar millor les dades a comparar, dels dos models, inicial i entrenat amb paràmetres (Accuracy + Log_Loss)

SVM_wine_comp = pd.DataFrame( data = [('SVM Incicial', accuracy_SVM_wine, Loss_SVM_wine), 
                                  ('SVM Paràmetres', accuracy_SVM_wine_2, Loss_SVM_wine_2)],
                         columns = ['Model', 'Accuracy', 'Log_Loss'])
SVM_wine_comp


Model  Accuracy  Log_Loss
0    SVM Incicial  0.805556  0.480189
1  SVM Paràmetres  0.833333  0.419474

In [26]:
# Generem un DF de la confusion matrix del model entrenat amb paràmetres

cm_SVM_wine_2_DF = pd.DataFrame(cm_SVM_wine_2, index = ['Class 1', 'Class 2', 'Class 3'],
                                 columns = ['Pred.SVM_2 Class 1', 'Pred.SVM_2 Class 2', 'Pred.SVM_2 Class 3'])

print('ConfusionMatrix model SVM entrenat amb paràmetres:')

cm_SVM_wine_2_DF


ConfusionMatrix model SVM entrenat amb paràmetres:


Pred.SVM_2 Class 1  Pred.SVM_2 Class 2  Pred.SVM_2 Class 3
Class 1                  14                   0                   0
Class 2                   0                  10                   4
Class 3                   1                   1                   6

In [27]:
# i la comparem amb la del model incial.

print('ConfusionMatrix model SVM inicial:')

cm_SVM_wine_DF

ConfusionMatrix model SVM inicial:


Pred.SVM Class 1  Pred.SVM Class 2  Pred.SVM Class 3
Class 1                14                 0                 0
Class 2                 0                11                 3
Class 3                 0                 4                 4

--------------------------------------------------------------------------------------------------------------------------------

### Exercici 4
 
Compara el seu rendiment fent servir l’aproximació traint/test o cross-validation.

- Comparem els diferents models generats amb cross-validation (models inicials i models entrenats)

In [28]:
# Predicció del 'score' de 'cross validation'

from sklearn.model_selection import cross_val_score

cv_DTC_wine = cross_val_score(estimator = DTC_wine, X = x, y = y, cv = 4)
cv_SVM_wine = cross_val_score(estimator = SVM_wine, X = x, y = y, cv = 4)
cv_DTC_wine_2 = cross_val_score(estimator = DTC_wine_2, X = x, y = y, cv = 4)
cv_SVM_wine_2 = cross_val_score(estimator = SVM_wine_2, X = x, y = y, cv = 4)


In [29]:
# Generem un DF per comparar els diferents valors obtinguts de la 'cross-validation'

cv_comp = pd.DataFrame( data = [('DTC Inicial', cv_DTC_wine.mean()),
                                ('DTC Paràmetres', cv_DTC_wine_2.mean()),
                                ('SVM Inicial', cv_SVM_wine.mean()),
                                ('SVM Paràmetres', cv_SVM_wine_2.mean())],
                         columns = ['Model', 'Cross-validation'])

cv_comp = cv_comp.sort_values(by=['Cross-validation'], ascending=False)

cv_comp

Model  Cross-validation
1  DTC Paràmetres          0.893813
0     DTC Inicial          0.859722
3  SVM Paràmetres          0.736237
2     SVM Inicial          0.663131

Comentaris:

- Havent realitzat 'cross-validation' dels dieferents models generats (inicials i amb paràmetres) podríem dir que el model mes ajustat es el __DecisionTreeClassifier__ __amb__ __paràmtres__ (DTC Paràmetres)

--------------------------------------------------------------------------------------------------------------------------------

### Exercici 5
   Aplica algun procés d'enginyeria per millorar els resultats (normalització, estandardització, mostreig...)

Aplicarem aquests processos al model que hem considerat més ajustat: __DecisionTreeClassifier__ __amb__ __paràmtres__ (DTC Paràmetres)

- __Normalització / Estandarització__: Transforma les dades per a que totes les variables tinguin un rang enre 0 i 1.

In [30]:
# Normalitzem les dades

from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
x_train_norm = scaler.fit_transform(x_train)
x_test_norm = scaler.transform(x_test)


In [31]:
# Entrenem el model

DTC_wine_2_norm = DTC_wine_2.fit(x_train_norm,y_train)
pred_DTC_wine_2_norm = DTC_wine_2_norm.predict(x_test_norm)

pred_DTC_wine_2_norm


array([1, 1, 3, 1, 2, 1, 2, 3, 2, 3, 2, 2, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2,
       2, 3, 3, 3, 2, 2, 2, 1, 1, 2, 3, 1, 1, 1], dtype=int64)

In [32]:
# Calculem 'Accuracy', log_loss' i fem 'Cross-validation'

accuracy_DTC_wine_2_norm = metrics.accuracy_score(y_test, pred_DTC_wine_2_norm)

print('accuracy:', accuracy_DTC_wine_2_norm)

Loss_DTC_wine_2_norm = log_loss(y_test, DTC_wine_2_norm.predict_proba(x_test_norm))

print('log_loss:', Loss_DTC_wine_2_norm)

cv_DTC_wine_2_norm = cross_val_score(estimator = DTC_wine_2_norm, X = x, y = y, cv = 4)

print('Cross-validation:', cv_DTC_wine_2_norm.mean())


accuracy: 0.9444444444444444
log_loss: 1.9188209108283725
Cross-validation: 0.9214646464646465


In [33]:
# Calculem la confusion matrix

cm_DTC_wine_2_norm = metrics.confusion_matrix(y_test, pred_DTC_wine_2_norm)

cm_DTC_wine_2_norm

# la presentem en DF

cm_DTC_wine_2_norm_DF = pd.DataFrame(cm_DTC_wine_2_norm, index = ['Class 1', 'Class 2', 'Class 3'], columns = ['Pred.NORM Class 1', 'Pred.NORM Class 2', 'Pred.NORM Class 3'])

cm_DTC_wine_2_norm_DF



Pred.NORM Class 1  Pred.NORM Class 2  Pred.NORM Class 3
Class 1                 13                  1                  0
Class 2                  0                 14                  0
Class 3                  0                  1                  7

--------------------------------------------------------------------------------------------------------------------------------

- __Selecció millors features__: Selecciona les millors features (característiques) de les dades.

In [34]:
# Seleccionem les 10 millors features

from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif  # evalua la relació entre cada variable i la variable objectiu

selector = SelectKBest(f_classif, k=10)  # K: nº de millors features que volem.
x_train_sel = selector.fit_transform(x_train, y_train)
x_test_sel = selector.transform(x_test)

In [35]:
# Entrenem el model

DTC_wine_2_sel = DTC_wine_2.fit(x_train_sel,y_train)
pred_DTC_wine_2_sel = DTC_wine_2_norm.predict(x_test_sel)

pred_DTC_wine_2_sel

array([1, 1, 3, 1, 2, 1, 2, 3, 2, 3, 1, 1, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2,
       2, 3, 3, 3, 2, 2, 2, 1, 1, 2, 3, 1, 1, 1], dtype=int64)

In [36]:
# Calculem 'Accuracy', log_loss' i fem 'Cross-validation'

accuracy_DTC_wine_2_sel = metrics.accuracy_score(y_test, pred_DTC_wine_2_sel)

print('accuracy:', accuracy_DTC_wine_2_sel)

Loss_DTC_wine_2_sel = log_loss(y_test, DTC_wine_2_sel.predict_proba(x_test_sel))

print('log_loss:', Loss_DTC_wine_2_sel)

cv_DTC_wine_2_sel = cross_val_score(estimator = DTC_wine_2_sel, X = x, y = y, cv = 4)

print('Cross-validation:', cv_DTC_wine_2_sel.mean())

accuracy: 0.9722222222222222
log_loss: 0.9594104554141872
Cross-validation: 0.9271464646464647


In [37]:
# Calculem la confusion matrix

cm_DTC_wine_2_sel = metrics.confusion_matrix(y_test, pred_DTC_wine_2_sel)

cm_DTC_wine_2_sel

# la presentem en DF

cm_DTC_wine_2_sel_DF = pd.DataFrame(cm_DTC_wine_2_sel, index = ['Class 1', 'Class 2', 'Class 3'], columns = ['Pred.SEL Class 1', 'Pred.SEL Class 2', 'Pred.SEL Class 3'])

cm_DTC_wine_2_sel_DF

Pred.SEL Class 1  Pred.SEL Class 2  Pred.SEL Class 3
Class 1                14                 0                 0
Class 2                 0                14                 0
Class 3                 1                 0                 7

-------------------------------------------------------------------------------------------------------------------------------

- __Mostreig__: Si al model hi ha més valors d'una classe que d'una altra, amb el mostreig podem aconseguir que el model no es decanti cap a la classe majoritària i donar més pes a la classe minoritària.

In [38]:
from imblearn.under_sampling import RandomUnderSampler

# amb 'random under-sampling' s'aconsegueix reduïr el nombre d'instànices de la 'classe' majoritària fins a un nivell igual al de la 'classe' minoritària.

# Definim l'estartgeia:
mostreig = RandomUnderSampler(sampling_strategy='majority', random_state=42)

# Apliquem 'under-sampling' al model d'entrenament:
x_train_mostreig, y_train_mostreig = mostreig.fit_resample(x_train, y_train)

x_train_mostreig, x_test_mostreig, y_train_mostreig, y_test_mostreig = train_test_split(x_train_mostreig, y_train_mostreig, test_size=0.2, random_state=42)




In [39]:
# Entrenem el model

DTC_wine_2_mostreig = DTC_wine_2.fit(x_train_mostreig, y_train_mostreig)
pred_DTC_wine_2_mostreig = DTC_wine_2_mostreig.predict(x_test_mostreig)

pred_DTC_wine_2_mostreig

array([2, 1, 1, 2, 2, 3, 2, 3, 3, 1, 1, 3, 1, 3, 1, 2, 1, 3, 1, 3, 2, 2,
       1, 1, 1], dtype=int64)

In [40]:
# Calculem 'Accuracy', log_loss' i fem 'Cross-validation'

accuracy_DTC_wine_2_mostreig = metrics.accuracy_score(y_test_mostreig, pred_DTC_wine_2_mostreig)

print('accuracy:', accuracy_DTC_wine_2_mostreig)

Loss_DTC_wine_2_mostreig = log_loss(y_test_mostreig, DTC_wine_2_mostreig.predict_proba(x_test_mostreig))

print('log_loss:', Loss_DTC_wine_2_mostreig)

cv_DTC_wine_2_mostreig = cross_val_score(estimator = DTC_wine_2_mostreig, X = x, y = y, cv = 4)

print('Cross-validation:', cv_DTC_wine_2_mostreig.mean())

accuracy: 0.96
log_loss: 1.3815510557964295
Cross-validation: 0.9272727272727272


In [41]:
# Calculem la confusion matrix

cm_DTC_wine_2_mostreig = metrics.confusion_matrix(y_test_mostreig, pred_DTC_wine_2_mostreig)

cm_DTC_wine_2_mostreig

# la presentem en DF

cm_DTC_wine_2_mostreig_DF = pd.DataFrame(cm_DTC_wine_2_mostreig, index = ['Class 1', 'Class 2', 'Class 3'], columns = ['Pred.MOSTREIG Class 1', 'Pred.MOSTREIG Class 2', 'Pred.MOSTREIG Class 3'])

cm_DTC_wine_2_mostreig_DF

Pred.MOSTREIG Class 1  Pred.MOSTREIG Class 2  Pred.MOSTREIG Class 3
Class 1                     11                      1                      0
Class 2                      0                      6                      0
Class 3                      0                      0                      7

--------------------------------------------------------------------------------------------------------------------------------

### RESUM

In [42]:
# Generem un DF per comparar les dades obtingudes en l'analisis dels diferents models

Comparatiu_Accuracy_df = pd.DataFrame(data = [('DecisionTreeClassifier:', accuracy_DTC_wine, Loss_DTC_wine, cv_DTC_wine.mean()),
                                              ('Support Vector Machines:', accuracy_SVM_wine, Loss_SVM_wine, cv_SVM_wine.mean()),
                                              ('DecisionTreeClassifier Entrenat:',accuracy_DTC_wine_2, Loss_DTC_wine_2, cv_DTC_wine_2.mean()),
                                              ('Support Vector Machines Entrenat:', accuracy_SVM_wine_2, Loss_SVM_wine_2, cv_SVM_wine_2.mean()),
                                              ('DecisionTreeClassifier Entrenat Normalitzat:', accuracy_DTC_wine_2_norm, Loss_DTC_wine_2_norm, cv_DTC_wine_2_norm.mean()),
                                              ('DecisionTreeClassifier Entrenat Selecció:',accuracy_DTC_wine_2_sel, Loss_DTC_wine_2_sel, cv_DTC_wine_2_sel.mean()),
                                              ('DecisionTreeClassifier Entrenat Mostreig:', accuracy_DTC_wine_2_mostreig, Loss_DTC_wine_2_mostreig, cv_DTC_wine_2_mostreig.mean())],
                                      columns= ['model', 'Accuracy', 'Log_Loss', 'Cross-Validation'])

Comparatiu_Accuracy_df = Comparatiu_Accuracy_df.sort_values(by=['Accuracy'], ascending=False)

Comparatiu_Accuracy_df

model  Accuracy  Log_Loss  \
5     DecisionTreeClassifier Entrenat Selecció:  0.972222  0.959410   
6     DecisionTreeClassifier Entrenat Mostreig:  0.960000  1.381551   
0                       DecisionTreeClassifier:  0.944444  1.918821   
2              DecisionTreeClassifier Entrenat:  0.944444  1.918821   
4  DecisionTreeClassifier Entrenat Normalitzat:  0.944444  1.918821   
3             Support Vector Machines Entrenat:  0.833333  0.419474   
1                      Support Vector Machines:  0.805556  0.480189   

   Cross-Validation  
5          0.927146  
6          0.927273  
0          0.859722  
2          0.893813  
4          0.921465  
3          0.736237  
1          0.663131

In [43]:
# Generem un DF per comparar totes les Confusion Matrix obtingudes

cm_DataFrames = [cm_DTC_wine_DF, cm_SVM_wine_DF, cm_DTC_wine_2_DF, cm_SVM_wine_2_DF, cm_DTC_wine_2_norm_DF, cm_DTC_wine_2_sel_DF, cm_DTC_wine_2_mostreig_DF]

cm_comparatiu = pd.concat(cm_DataFrames, axis = 1)

cm_comparatiu



Pred.DTC Class 1  Pred.DTC Class 2  Pred.DTC Class 3  \
Class 1                13                 1                 0   
Class 2                 0                14                 0   
Class 3                 1                 0                 7   

         Pred.SVM Class 1  Pred.SVM Class 2  Pred.SVM Class 3  \
Class 1                14                 0                 0   
Class 2                 0                11                 3   
Class 3                 0                 4                 4   

         Pred.DTC_2 Class 1  Pred.DTC_2 Class 2  Pred.DTC_2 Class 3  \
Class 1                  13                   1                   0   
Class 2                   0                  14                   0   
Class 3                   0                   1                   7   

         Pred.SVM_2 Class 1  ...  Pred.SVM_2 Class 3  Pred.NORM Class 1  \
Class 1                  14  ...                   0                 13   
Class 2                   0  ...                   4                  0   
Class 3                   1  ...                   6                  0   

         Pred.NORM Class 2  Pred.NORM Class 3  Pred.SEL Class 1  \
Class 1                  1                  0                14   
Class 2                 14                  0                 0   
Class 3                  1                  7                 1   

         Pred.SEL Class 2  Pred.SEL Class 3  Pred.MOSTREIG Class 1  \
Class 1                 0                 0                     11   
Class 2                14                 0                      0   
Class 3                 0                 7                      0   

         Pred.MOSTREIG Class 2  Pred.MOSTREIG Class 3  
Class 1                      1                      0  
Class 2                      6                      0  
Class 3                      0                      7  

[3 rows x 21 columns]

-------------------------------------------------------------------------------------------------------------------------------

#### Comentaris RESUM

Havent realitzat i analitzat els models de classificació 'DecisionTreeClassifier' (DTC) i 'SupportVectorMachines' (SVM) hem obtingut els següents resultats:

- Models Inicials:
  - Accuracy: El 'DTC' inicial obté una 'accuracy' (percentatge de prediccions correctes) del 94% per sobre de 80% de la 'accuracy'obtinguda amb el model 'SVM'. Analitzant la 'accuracy' podriem dir que el model més ajustat és el DTC.
  - Confusion Matrix: El 'DTC' inicial acumula 2 errors en les prediccions respecte els 7 errors del model 'SVM'. Com en el cas de l'analisi de l'accurcy, podriem dir que analitzant la confusion matrix, el model 'DTC' inicial és més precís.
  - Logistic Loss: En aquest cas, al 'DTC' inicial obtenim una 'log_loss' superior a la obtinguda amb el model 'SVM' inicial. En aques cas, podríem dir que les prediccions del model 'SVM' inicial podríen ser millors respecte a les del model 'DTC' inicial.
  - Cross validation: En aquest cas el 'DTC' inicial obté un valor de CV molt superior al obtingut al model 'SVM' incial. Podríem dir que en aquest cas el model 'DTC' inicial és més precís.
        
- Models Entrenats:
    - Entrenant els models, el model 'DTC' només obté una petita millora en el Cross Validation, mentre que el model 'SVM' obté millores en els tots els analisis, inclús en la confusion matrix que pasa del 7 errors als 6 errors.
    - Malgrat no haver obtingut millores substancials després d'entrenar el model 'DTC' inicial, el model DTC entrenat a priori continua sent més ajustat que el model 'SVM' entrenat.
        
- Millora dels rsultats:
    - Hem realitzat processos de millora sobre el model 'DTC' entrenat
     - Normalització: en principi no es millora el model entrenat
     - Selecció millors 'features': en aquest cas s'aconsegueix aumentar el percentatge 'accuracy' fins al 97% i reduir substancialment la 'log_loss'. Malgrat la CV es manté, s'aconsegueix reduir a 1 error les predccions a la confusion matrix.
     - Mostreig: s'aconsegueix aumnetar sensiblement l'accuracy i la CV, i es redueix en menor quantia la log_loss. S'aconsegueix reduir també els errors de predicció de la confusion matrix, reduint també a 1 error.
        

#### Conclusió

Havent realitzat, entrenat, millorat i analitzat els diferents models, es considera que possiblement el millor model per realitzar les prediccions d'indentificació de les diferents Classes és el model __DecisionTreeClassifier Entrenat Selecció__

- __Model DTC entrenat amb paràmetres i millorat amb la selecció de les millors 'features'__